File: cpp/libs/opendnp3/src/opendnp3/app/MeasurementTypes.cpp
/**
 * Licensed to Green Energy Corp (www.greenenergycorp.com) under one or
 * more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Green Energy Corp licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except in
 * compliance with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This project was forked on 01/01/2013 by Automatak, LLC and modifications
 * may have been made to this file. Automatak, LLC licenses these modifications
 * to you under the terms of the License.
 */

#include "MeasurementTypes.h"

namespace opendnp3
{

namespace quality {	

	bool GetBinaryValue(uint8_t quality)
	{
		return (quality & static_cast<uint8_t>(BinaryQuality::STATE)) != 0;
	}

	uint8_t GetBinaryQuality(uint8_t q, bool value)
	{
		return (value) ? (q | static_cast<uint8_t>(BinaryQuality::STATE)) : (q & (~static_cast<uint8_t>(BinaryQuality::STATE)));
	}

}


// ------------ Binary ---------------

Binary::Binary() : TypedMeasurement(false, quality::RESTART)
{}

Binary::Binary(bool value) : TypedMeasurement(value, quality::GetBinaryQuality(quality::ONLINE, value))
{}

Binary::Binary(uint8_t quality) : TypedMeasurement(quality::GetBinaryValue(quality), quality)
{}

Binary::Binary(uint8_t quality, DNPTime time) : TypedMeasurement(quality::GetBinaryValue(quality), quality, time)
{}

Binary::Binary(bool value, uint8_t quality) : TypedMeasurement(value, quality::GetBinaryQuality(quality, value))
{}

Binary::Binary(bool value, uint8_t quality, DNPTime time) : TypedMeasurement(value, quality::GetBinaryQuality(quality, value), time)
{}

bool Binary::IsQualityOnlineOnly() const
{
	bool isOnline = quality & static_cast<uint8_t>(BinaryQuality::ONLINE);
	uint8_t allowedBits = static_cast<uint8_t>(BinaryQuality::ONLINE) | static_cast<uint8_t>(BinaryQuality::STATE);
	bool otherBitsClear = (quality & ~allowedBits) == 0;

	return isOnline && otherBitsClear;	
}

bool Binary::IsEvent(const Binary& newValue) const
{
	return quality != newValue.quality;
}


// ------------ Double Bit Binary ---------------

DoubleBitBinary::DoubleBitBinary() : TypedMeasurement(DoubleBit::INDETERMINATE, quality::RESTART)
{}

DoubleBitBinary::DoubleBitBinary(DoubleBit value) : TypedMeasurement(value, GetQual(quality::ONLINE, value))
{}

DoubleBitBinary::DoubleBitBinary(uint8_t quality) : TypedMeasurement(GetValue(quality), quality)
{}

DoubleBitBinary::DoubleBitBinary(uint8_t quality, DNPTime time) : TypedMeasurement(GetValue(quality), quality, time)
{}

DoubleBitBinary::DoubleBitBinary(DoubleBit value, uint8_t quality) : TypedMeasurement(value, GetQual(quality, value))
{}

DoubleBitBinary::DoubleBitBinary(DoubleBit value, uint8_t quality, DNPTime time) : TypedMeasurement(value, GetQual(quality, value), time)
{}

bool DoubleBitBinary::IsEvent(const DoubleBitBinary& newValue) const
{
	return quality != newValue.quality;
}

DoubleBit DoubleBitBinary::GetValue(uint8_t quality)
{
	// the value is the top 2 bits, so down-shift 6
	uint8_t value = quality >> 6;
	return DoubleBitFromType(value);
}

uint8_t DoubleBitBinary::GetQual(uint8_t quality, DoubleBit state)
{
	uint8_t value = DoubleBitToType(state) << 6;
	return (QualityMask & quality) | value;
}

// ------------ Binary Output Status ---------------


BinaryOutputStatus::BinaryOutputStatus() : TypedMeasurement(false, quality::RESTART)
{}

BinaryOutputStatus::BinaryOutputStatus(bool value) : TypedMeasurement(value, quality::GetBinaryQuality(quality::ONLINE, value))
{}

BinaryOutputStatus::BinaryOutputStatus(uint8_t quality) : TypedMeasurement(quality::GetBinaryValue(quality), quality)
{}

BinaryOutputStatus::BinaryOutputStatus(uint8_t quality, DNPTime time) : TypedMeasurement(quality::GetBinaryValue(quality), quality, time)
{}

BinaryOutputStatus::BinaryOutputStatus(bool value, uint8_t quality) : TypedMeasurement(value, quality::GetBinaryQuality(quality, value))
{}

BinaryOutputStatus::BinaryOutputStatus(bool value, uint8_t quality, DNPTime time) : TypedMeasurement(value, quality::GetBinaryQuality(quality, value), time)
{}

bool BinaryOutputStatus::IsEvent(const BinaryOutputStatus& newValue) const
{
	return quality != newValue.quality;
}

// ------------ Analog ---------------

Analog::Analog() : TypedMeasurement(quality::RESTART)
{}

Analog::Analog(double value) : TypedMeasurement(value, quality::ONLINE)
{}

Analog::Analog(double value, uint8_t quality) : TypedMeasurement(value, quality)
{}

Analog::Analog(double value, uint8_t quality, DNPTime time) : TypedMeasurement<double>(value, quality, time)
{}

bool Analog::IsEvent(const Analog& newValue, double deadband) const
{
	return measurements::IsEvent(newValue, *this, deadband);
}

// ------------ Counter ---------------



Counter::Counter() : TypedMeasurement(0, quality::RESTART) {}

Counter::Counter(uint32_t value) : TypedMeasurement<uint32_t>(value, quality::ONLINE)
{}

Counter::Counter(uint32_t value, uint8_t quality) : TypedMeasurement<uint32_t>(value, quality)
{}

Counter::Counter(uint32_t value, uint8_t quality, DNPTime time) : TypedMeasurement<uint32_t>(value, quality, time)
{}

bool Counter::IsEvent(const Counter& newValue, uint32_t aDeadband) const
{
	if(quality != newValue.quality) return true;
	else
	{
		return measurements::IsEvent<uint32_t, uint64_t>(this->value, newValue.value, aDeadband);
	}
}

// ------------ Frozen Counter ---------------


FrozenCounter::FrozenCounter() : TypedMeasurement(0, quality::RESTART) {}

FrozenCounter::FrozenCounter(uint32_t value) : TypedMeasurement<uint32_t>(value, quality::ONLINE)
{}

FrozenCounter::FrozenCounter(uint32_t value, uint8_t quality) : TypedMeasurement<uint32_t>(value, quality)
{}

FrozenCounter::FrozenCounter(uint32_t value, uint8_t quality, DNPTime time) : TypedMeasurement<uint32_t>(value, quality, time)
{}

bool FrozenCounter::IsEvent(const FrozenCounter& newValue, uint32_t aDeadband) const
{
	if(quality != newValue.quality) return true;
	else
	{
		return measurements::IsEvent<uint32_t, uint64_t>(this->value, newValue.value, aDeadband);
	}
}

// ------------ Analog Output Status ---------------



AnalogOutputStatus::AnalogOutputStatus() : TypedMeasurement<double>(quality::RESTART) {}


AnalogOutputStatus::AnalogOutputStatus(double value) : TypedMeasurement<double>(value, quality::ONLINE)
{}

AnalogOutputStatus::AnalogOutputStatus(double value, uint8_t quality) : TypedMeasurement<double>(value, quality)
{}

AnalogOutputStatus::AnalogOutputStatus(double value, uint8_t quality, DNPTime time) : TypedMeasurement<double>(value, quality, time)
{}

bool AnalogOutputStatus::IsEvent(const AnalogOutputStatus& newValue, double deadband) const
{
	return measurements::IsEvent(newValue, *this, deadband);
}


} // end ns

File: dotnet/CLRInterface/src/MeasurementTypes.cs
// Copyright 2013-2020 Automatak, LLC
//
// Licensed to Green Energy Corp (www.greenenergycorp.com) and Automatak
// LLC (www.automatak.com) under one or more contributor license agreements. 
// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership. Green Energy Corp and Automatak LLC license
// this file to you under the Apache License, Version 2.0 (the "License"); you
// may not use this file except in compliance with the License. You may obtain
// a copy of the License at:
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Automatak.DNP3.Interface
{
    /// <summary>
    /// A boolean measurement type (i.e. whether a switch is open/closed)
    /// </summary>
    public class Binary : TypedMeasurementBase<bool> 
    {		
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        /// <param name="time">timestamp</param>
	    public Binary(bool value, Flags quality, DNPTime time) :
            base(value, quality, time)
	    {}

        /// <summary>
        /// Constructor without a timestamp
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        public Binary(bool value, Flags quality) :
            base(value, quality)
        {}

        /// <summary>
        /// Constructor with only quality
        /// </summary>
        /// <param name="quality"></param>
        public Binary(Flags quality) : base(quality.ToBinaryValue(), quality)
        { }       
    }

    /// <summary>
    /// A boolean measurement type (i.e. whether a switch is open/closed)
    /// </summary>
    public class DoubleBitBinary : TypedMeasurementBase<DoubleBit> 
    {		
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        /// <param name="time">timestamp</param>
	    public DoubleBitBinary(DoubleBit value, Flags quality, DNPTime time) : 
            base(value, quality, time)
	    {}

        /// <summary>
        /// Constructor without a timestamp
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        public DoubleBitBinary(DoubleBit value, Flags quality) :
            base(value, quality)
        {}

        /// <summary>
        /// Constructor with only a quality
        /// </summary>
        /// <param name="value"></param>
        /// <param name="quality"></param>
        /// <param name="time"></param>
        public DoubleBitBinary(Flags quality)
            : base(quality.ToDoubleBitValue(), quality)
        { }
    }

    /// <summary>
    /// A discrete or continuous analog value
    /// </summary>
    public class Analog : TypedMeasurementBase<double>
    {
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        /// <param name="time">timestamp</param>
        public Analog(double value, Flags quality, DNPTime time)
            : base(value, quality, time)
	    {}

        /// <summary>
        /// Constructor without a timestamp
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        public Analog(double value, Flags quality) : base(value, quality)
        {}

        /// <summary>
        /// Constructor with only quality
        /// </summary>
        /// <param name="quality"></param>
        public Analog(Flags quality) : base(0.0, quality)
        { }
    }

    /// <summary>
    /// An increasing count of some event
    /// </summary>
    public class Counter : TypedMeasurementBase<System.UInt32>
    {

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        /// <param name="time">timestamp</param>
        public Counter(System.UInt32 value, Flags quality, DNPTime time): base(value, quality, time)
	    {}

        /// <summary>
        /// Constructor without a timestamp
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        public Counter(System.UInt32 value, Flags quality) : base(value, quality)
        {}

        /// <summary>
        /// Constructor with only quality
        /// </summary>
        /// <param name="quality"></param>
        public Counter(Flags quality) : base(0, quality)
        { }
    }

    /// <summary>
    /// An increasing count of some event frozen in time
    /// </summary>
    public class FrozenCounter : TypedMeasurementBase<System.UInt32>
    {

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        /// <param name="time">timestamp</param>
        public FrozenCounter(System.UInt32 value, Flags quality, DNPTime time): base(value, quality, time)
	    {}

        /// <summary>
        /// Constructor without a timestamp
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        public FrozenCounter(System.UInt32 value, Flags quality) : base(value, quality)
        {}

        /// <summary>
        /// Constructor with only quality
        /// </summary>
        /// <param name="quality"></param>
        public FrozenCounter(Flags quality) : base(0, quality)
        { }
    }

    /// <summary>
    /// Represents the state of a digital output
    /// </summary>
    public class BinaryOutputStatus : TypedMeasurementBase<bool>
    {

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        /// <param name="time">timestamp</param>
	    public BinaryOutputStatus(bool value, Flags quality, DNPTime time) :
            base(value, quality, time)
	    {}

        /// <summary>
        /// Constructor without a timestamp
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        public BinaryOutputStatus(bool value, Flags quality) :
            base(value, quality)
        {}

        /// <summary>
        /// Constructor with only quality
        /// </summary>
        /// <param name="quality"></param>
        public BinaryOutputStatus(Flags quality)
            : base(quality.ToBinaryValue(), quality)
        { }
    };

    /// <summary>
    /// Represents the status of analog output on an outstation.
    /// </summary>    
    public class AnalogOutputStatus : TypedMeasurementBase<double>
    {

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        /// <param name="time">timestamp</param>
        public AnalogOutputStatus(double value, Flags quality, DNPTime time) : base(value, quality, time)
	    {}

        /// <summary>
        /// Constructor without a timestamp
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        public AnalogOutputStatus(double value, Flags quality): base(value, quality)
        {}

         /// <summary>
        /// Constructor with only quality
        /// </summary>
        /// <param name="quality"></param>
        public AnalogOutputStatus(Flags quality) : base(0.0, quality)
        { }
    }

    public class OctetData
    {
        public OctetData(byte[] bytes)
        {
            if (bytes.Length < 1 || bytes.Length > 255)
            {
                throw new ArgumentException("byte array length must be in interval [1,255]", "bytes");
            }
            this.bytes = bytes;
        }

        public String AsString()
        {
            return new String(Encoding.ASCII.GetChars(bytes));
        }

        public byte[] Bytes
        {
            get { return bytes; }
        }

        private readonly byte[] bytes;    
    }

    public class OctetString : OctetData
    {
        public OctetString(byte[] bytes) : base(bytes)
        { }
    }

    public class TimeAndInterval
    {
        public TimeAndInterval(UInt64 time, UInt32 interval, Byte units)
        {
            this.time = time;
            this.interval = interval;
            this.units = units;
        }

        public TimeAndInterval(UInt64 time, UInt32 interval, IntervalUnits units)
            : this(time, interval, (Byte)units)
        {}

        IntervalUnits GetUnitsEnum()
        {
            if (Enum.IsDefined(typeof(IntervalUnits), units))
            {
                return (IntervalUnits)Enum.ToObject(typeof(IntervalUnits), units);
            }
            else
            {
                return IntervalUnits.Undefined;
            }
        }

        public override string ToString()
        {
            return String.Format("time: {0} interval: {1} units: {2}", time, interval, units);
        }

        public readonly UInt64 time;
        public readonly UInt32 interval;
        public readonly Byte units;	
    }

    /// <summary>
    /// Represents the status of analog output on an outstation.
    /// </summary>    
    public class SecurityStat : TypedMeasurementBase<System.UInt32>
    {

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="value">value of the measurement</param>
        /// <param name="quality">quality enumeration as a bitfield</param>
        /// <param name="time">timestamp</param>
        public SecurityStat(uint value, ushort assocId, Flags quality, DNPTime time)
            : base(value, quality, time)
        {
            this.assocId = assocId;
        }        

        UInt16 AssociationId
        {
            get
            {
                return assocId;
            }
        }

        private readonly UInt16 assocId;
    }

    

    public class BinaryCommandEvent
    {
        public BinaryCommandEvent(bool value, CommandStatus status, DNPTime time)
        {
            this.value = value;
            this.status = status;
            this.time = time;
        }

        public override string ToString()
        {
            return String.Format("value: {0} status: {1} time: {2}", value, status, time);
        }

        private readonly bool value;
        private readonly CommandStatus status;
        private readonly DNPTime time;

        public bool Value { get { return this.value; } }
        public CommandStatus Status { get { return this.status; } }
        public DNPTime Timestamp { get { return this.time; } }
    }

    public class AnalogCommandEvent
    {
        public AnalogCommandEvent(double value, CommandStatus status, DNPTime time)
        {
            this.value = value;
            this.status = status;
            this.time = time;
        }

        public override string ToString()
        {
            return String.Format("value: {0} status: {1} time: {2}", value, status, time);
        }

        private readonly double value;
        private readonly CommandStatus status;
        private readonly DNPTime time;

        public double Value { get { return this.value; } }
        public CommandStatus Status { get { return this.status; } }
        public DNPTime Timestamp { get { return this.time; } }
    }
}
